# In this project I will work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

# The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
# The original dataset isn't available on Kaggle anymore, but you can find it here. https://data.world/data-society/used-cars-data

# The aim is to clean the data and analyze the included user car listing.

In [37]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [38]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [39]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [40]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### We can see that some of the data has null values, with which we have to do something. Also they use camelcase instead of snakecase which is preferred in Python, so I would need to convert them.

In [41]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [42]:
autos.rename(columns={'yearOfRegistration': 'registration_year', 
                      'monthOfRegistration': 'registration_month', 
                      'notRepairedDamage': 'unrepaired_damage', 
                      'dateCreated': 'ad_created',
                      'dateCrawled': 'date_crawled',
                     'offerType': 'offer_type',
                     'vehicleType': 'vehicle_type',
                     'powerPS': 'power_PS',
                     'fuelType': 'fuel_type',
                     'nrOfPictures':'nr_of_pictures',
                     'postalCode':'postal_code',
                     'lastSeen' : 'last_seen'}, inplace=True)

### I have changed the column names to be snake case and renamed some of the columns to make more sense and so it would be conventional and easier to query the df

In [43]:
autos.describe()

registration_year      power_PS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

### Price and odometer columns are both numberic values but they are stored as text files

In [44]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object

In [45]:
autos['price'] = autos['price'].str.replace('$' , '')
autos['price'] = autos['price'].str.replace(',' , '')
autos['price'] = autos['price'].astype(int)

In [46]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [47]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [48]:
autos['odometer'] = autos['odometer'].str.replace('km' , '')
autos['odometer'] = autos['odometer'].str.replace(',' , '')
autos['odometer'] = autos['odometer'] .astype(int)

In [49]:
autos['odometer']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: int64

In [50]:
autos.rename(columns = {'odometer' : 'odometer_km', 'price' : 'price_dollars'}, inplace = True)

### Will start looking into columns to ensure that data is correct and doesn't have any unrealistically high or low outliers

### Firstly odometer_km

In [51]:
autos['odometer_km'].unique()


array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [52]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [53]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

### Odometer_km seems to be ok

### Price_dollars column

In [54]:
autos['price_dollars'].unique().shape

(2357,)

In [55]:
autos['price_dollars'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [56]:
autos['price_dollars'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_dollars, dtype: int64

In [57]:
autos['price_dollars'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_dollars, dtype: int64

In [58]:
autos.loc[autos['price_dollars'] >= 1000000]

date_crawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type  price_dollars   abtest vehicle_type  \
2897   privat    Angebot       11111111     test    limousine   
7814   privat    Angebot        1300000  control        coupe   
11137  privat    Angebot       10000000  control        coupe   
22947  privat    Angebot        1234566  control        kombi   
24384  privat    Angebot       11111111     test          NaN   
27371  privat    Angebot       12345678  control          NaN   
39377  privat    Angebot       12345678  control          NaN   
39705  privat    Angebot       99999999  control    limousine   
42221  privat    Angebot       27322222  control    limousine   
47598  privat    Angebot       12345678  control    limousine   
47634  privat    Angebot        3890000     test        coupe   

       registration_year    gearbox  power_PS     model  odometer_km  \
2897                1973    manuell        48    escort        50000   
7814                1992        NaN         0       NaN        50000   
11137               1960    manuell       368       NaN       100000   
22947               1999  automatik       190       NaN       150000   
24384               1995        NaN         0       NaN       150000   
27371               2017        NaN        95     punto       150000   
39377               2018    manuell        95       v40       150000   
39705               1999  automatik       224  s_klasse       150000   
42221               2014    manuell       163        c4        40000   
47598               2001    manuell       101    vectra       150000   
47634               2006        NaN       799       NaN         5000   

       registration_month fuel_type           brand unrepaired_damage  \
2897                    3    benzin            ford              nein   
7814                   12       NaN  sonstige_autos              nein   
11137                   1    benzin  sonstige_autos              nein   
22947                   2    diesel             bmw               NaN   
24384                   0       NaN      volkswagen               NaN   
27371                   0       NaN            fiat               NaN   
39377                   6       NaN           volvo              nein   
39705                   9    benzin   mercedes_benz               NaN   
42221                   2    diesel         citroen               NaN   
47598                   3    benzin            opel              nein   
47634                   7       NaN  sonstige_autos              nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
2897   2016-03-12 00:00:00               0        94469  2016-03-12 22:45:27  
7814   2016-04-04 00:00:00               0        60598  2016-04-05 11:34:11  
11137  2016-03-29 00:00:00               0        73033  2016-04-06 21:18:11  
22947  2016-03-22 00:00:00               0        17454  2016-04-02 03:17:32  
24384  2016-03-21 00:00:00               0        18519  2016-03-21 14:40:18  
27371  2016-03-09 00:00:00               0  

In [59]:
autos.loc[autos['registration_year'] < 1920]

date_crawled                                              name  \
3679   2016-04-04 00:36:17                                        Suche_Auto   
10556  2016-04-01 06:02:10                                        UNFAL_Auto   
15898  2016-03-08 10:50:05       Tausch_alles_aus_meinen_Anzeigen_gegen_Auto   
22316  2016-03-29 16:56:41             VW_Kaefer.__Zwei_zum_Preis_von_einem.   
22659  2016-03-14 08:51:18                                      Opel_Corsa_B   
24511  2016-03-17 19:45:11                       Trabant__wartburg__Ostalgie   
28693  2016-03-22 17:48:41                                    Renault_Twingo   
30781  2016-03-25 13:47:46                  Opel_Calibra_V6_DTM_Bausatz_1:24   
32585  2016-04-02 16:56:39                                        UNFAL_Auto   
33295  2016-03-12 10:49:42            Ich_verkaufe/suche_fuer_sie_ihr_Auto!!   
35238  2016-03-26 13:45:20  Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima   
42181  2016-03-27 19:50:53         SAMSUNG_55“_3D_Tv_und_Soundbar_gegen_Auto   
45157  2016-03-11 22:37:01                                        Motorhaube   
46213  2016-04-02 13:47:16                                     Bellier_Vario   
49283  2016-03-15 18:38:53                                        Citroen_HY   

       seller offer_type  price_dollars   abtest vehicle_type  \
3679   privat    Angebot              1     test          NaN   
10556  privat    Angebot            450  control          NaN   
15898  privat    Angebot              0     test          NaN   
22316  privat    Angebot           1500  control          NaN   
22659  privat    Angebot            500     test          NaN   
24511  privat    Angebot            490  control          NaN   
28693  privat    Angebot            599  control   kleinwagen   
30781  privat    Angebot             30     test          NaN   
32585  privat    Angebot            450  control          NaN   
33295  privat    Angebot              0  control          NaN   
35238  privat    Angebot              0  control          NaN   
42181  privat    Angebot              0     test          NaN   
45157  privat    Angebot             15  control          NaN   
46213  privat    Angebot              0     test   kleinwagen   
49283  privat    Angebot           7750  control          NaN   

       registration_year  gearbox  power_PS    model  odometer_km  \
3679                1910      NaN         0      NaN         5000   
10556               1800      NaN      1800      NaN         5000   
15898               1910      NaN         0      NaN         5000   
22316               1000  manuell         0   kaefer         5000   
22659               1910      NaN         0    corsa       150000   
24511               1111      NaN         0      NaN         5000   
28693               1910  manuell         0      NaN         5000   
30781               1910      NaN         0  calibra       100000   
32585               1800      NaN      1800      NaN         5000   
33295               1910      NaN         0      NaN         5000   
35238               1500      NaN         0      NaN         5000   
42181               1910      NaN         0      NaN         5000   
45157               1910      NaN         0      NaN         5000   
46213               1910      NaN         0      NaN         5000   
49283               1001      NaN         0   andere         5000   

       registration_month fuel_type           brand unrepaired_damage  \
3679                    0       NaN  sonstige_autos               NaN   
10556                   2       NaN      mitsubishi              nein   
15898                   0       NaN  sonstige_autos               NaN   
22316                   0    benzin      volkswagen               NaN   
22659                   0       NaN            opel               NaN   
24511                   0       NaN         trabant               NaN   
28693                   0    benzin         renault               NaN   
307

In [60]:
autos.loc[autos['registration_year'] > 2018]

date_crawled                                               name  \
453    2016-03-28 13:51:12                                         Armee_Jeep   
4164   2016-03-29 18:39:40  Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...   
4549   2016-04-01 21:57:05                                         Kompressor   
5763   2016-03-05 19:54:49             Suche_Auspuff_Anlage_honda_crx_del_sol   
6308   2016-03-12 17:38:17                              Kaufe_Autos_jeglicher   
8012   2016-03-23 16:43:29                      Opel_GT_Karosserie_mit_Brief!   
8360   2016-03-11 22:56:30                           Vito_touret_119_Blue_Tec   
13559  2016-03-19 15:57:44          Saab_9000_CSE_Automatik_2_3_ltr._mit_EGSD   
14341  2016-03-23 01:36:20                                  Hole_kostenlos_ab   
22799  2016-03-20 18:56:44                                  Subaru_Impreza_GT   
24519  2016-03-05 17:53:37  4x_Winterreifen_auf_Alufelge_der_naechste_Wint...   
25003  2016-03-06 16:06:20              Reo_Vorkriegs_Oldtimer_Rennwagen_1928   
27578  2016-03-31 22:49:04                                   VW_GOLF_2_SYNCRO   
27618  2016-03-29 22:39:32  Golf_1_75ps_5911km_Original_Automatik_einer_de...   
33950  2016-03-23 21:52:25  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...   
38076  2016-04-04 22:54:47                                 Mercedes_Benz_A180   
38342  2016-03-07 18:52:22              Suche_Auto_fuer_die_jagd_zum_tauschen   
42079  2016-03-17 19:55:29                      APE_50___deutsche_Papiere_!!!   
49153  2016-03-12 01:36:59                                        Corsa_c20xe   
49185  2016-03-28 11:45:31              Alfa_Romeo_156_Sportwagon_1_8_T_Spark   
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   

       seller offer_type  price_dollars   abtest vehicle_type  \
453    privat    Angebot           9800     test          NaN   
4164   privat    Angebot             49  control          NaN   
4549   privat    Angebot           1600     test          NaN   
5763   privat    Angebot            100     test          NaN   
6308   privat    Angebot              0     test          NaN   
8012   privat    Angebot            700     test          NaN   
8360   privat    Angebot          42800  control          NaN   
13559  privat    Angebot              0  control          NaN   
14341  privat    Angebot              0     test          NaN   
22799  privat    Angebot           9000     test          NaN   
24519  privat    Angebot            250     test          NaN   
25003  privat    Angebot           6500  control          NaN   
27578  privat    Angebot           1700     test          NaN   
27618  privat    Angebot           7999     test          NaN   
33950  privat    Angebot           7999     test          NaN   
38076  privat    Angebot          18000     test          NaN   
38342  privat    Angebot              0  control          NaN   
42079  privat    Angebot            500  control          NaN   
49153  privat    Angebot           2500     test          NaN   
49185  privat    Angebot           1200  control          NaN   
49910  privat    Angebot          22200     test          NaN   

       registration_year    gearbox  power_PS     model  odometer_km  \
453                 4500    manuell         0    andere         5000   
4164                5000        NaN         0      golf         5000   
4549                4100        NaN         0       NaN         5000   
5763                2019    manuell       125  cr_reihe        20000   
6308                9996        NaN         0       NaN        10000   
8012                9999        NaN         0    andere        10000   
8360                6200  automatik         0      vito        10000   
13559               9000  automatik       170      9000       150000   
14341               9999        NaN         0       NaN        10000   
22799               5000    manuell       420   impreza         5000   
24519   

### We  can see that there are also invalid years and we shall remove those as well, all cars that have a registration_year over 2018

In [61]:
autos.loc[autos['price_dollars'] <= 123]

date_crawled                                               name  \
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
64     2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
...                    ...                                                ...   
49884  2016-03-11 13:55:30               Audi_a6_2.5l__Schnaeppchen_nur_heute   
49943  2016-03-16 20:46:08                                         Opel_astra   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   

       seller offer_type  price_dollars   abtest vehicle_type  \
25     privat    Angebot             90  control        kombi   
27     privat    Angebot              0  control          NaN   
30     privat    Angebot             80     test   kleinwagen   
55     privat    Angebot              1     test          NaN   
64     privat    Angebot             40     test          NaN   
...       ...        ...            ...      ...          ...   
49884  privat    Angebot              0     test        kombi   
49943  privat    Angebot              0  control          NaN   
49960  privat    Angebot              0  control   kleinwagen   
49974  privat    Angebot              0  control       cabrio   
49984  privat    Angebot              0     test          NaN   

       registration_year    gearbox  power_PS     model  odometer_km  \
25                  1996    manuell       116       NaN       150000   
27                  2005        NaN         0       NaN       150000   
30                  2002    manuell        60   2_reihe       150000   
55                  2017  automatik       224  e_klasse       125000   
64                  2011        NaN         0       5er       150000   
...                  ...        ...       ...       ...          ...   
49884               1999    manuell       150        a6       150000   
49943               2016    manuell       101     astra       150000   
49960               1999    manuell        60        ka       150000   
49974               1983    manuell        70      golf       150000   
49984               2000        NaN         0       NaN       150000   

       registration_month fuel_type           brand unrepaired_damage  \
25                      4    benzin            ford                ja   
27                      0       NaN            ford               NaN   
30                      6    benzin         peugeot                ja   
55                      7    benzin   mercedes_benz              nein   
64                      5       NaN             bmw               NaN   
...                   ...       ...             ...               ...   
49884                  11    diesel            audi               NaN   
49943                   8    benzin            opel               NaN   
49960                   6    benzin            ford               NaN   
49974                   2    benzin      volkswagen              nein   
49984                   0       NaN  sonstige_autos               NaN   

                ad_created  nr_of_pictures  postal_code            last_seen  
25     2016-03-21 00:00:00               0        27574  2016-04-01 05:16:49  
27     2016-03-27 00:00:00               0        66701  2016-03-27 18:45:01  
30     2016-03-14 00:00:00               0        57076  2016-03-14 11:47:31  
55     2016-03-06 00:00:00               0        22111  2016-03-08 05:45:44  
64     2016-04-05 00:00:00               0        40591  2016-04-07 12:16:01  
...                    ...             ...  

In [62]:
filter_prices_dollars = ~(~(autos["registration_month"].between(1,12)) | ~(autos["registration_year"].between(1910,2016)) | (autos["price_dollars"] <= 150) | ((autos["price_dollars"] >= 1000000) & ~((autos['price_dollars'] == 1300000)
                                         | (autos['price_dollars'] == 3890000))))

In [63]:
autos = autos[filter_prices_dollars]

### I have now removed all cars that are less than 150 or over 1 000 000 dollars except for 2 Ferraris that had a correct price.

### Moving onto date values, there are 5 columns that represent date values, and first three are represented as strings and last two as numeric values

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [64]:
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [65]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025018
2016-03-06    0.014310
2016-03-07    0.036367
2016-03-08    0.033168
2016-03-09    0.033074
2016-03-10    0.032315
2016-03-11    0.033074
2016-03-12    0.036912
2016-03-13    0.015731
2016-03-14    0.036580
2016-03-15    0.033784
2016-03-16    0.029283
2016-03-17    0.031462
2016-03-18    0.012817
2016-03-19    0.034211
2016-03-20    0.038167
2016-03-21    0.037243
2016-03-22    0.032505
2016-03-23    0.032458
2016-03-24    0.029164
2016-03-25    0.031036
2016-03-26    0.032244
2016-03-27    0.030989
2016-03-28    0.034827
2016-03-29    0.033974
2016-03-30    0.033524
2016-03-31    0.031818
2016-04-01    0.034400
2016-04-02    0.036011
2016-04-03    0.039328
2016-04-04    0.036604
2016-04-05    0.013078
2016-04-06    0.003104
2016-04-07    0.001421
Name: date_crawled, dtype: float64

### We can see that mostly all dates follow the same frequency except for 2016-04-06 2016-04-07, but that is not really importantt

In [66]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001114
2016-03-06    0.003933
2016-03-07    0.005402
2016-03-08    0.007131
2016-03-09    0.009642
2016-03-10    0.010495
2016-03-11    0.012296
2016-03-12    0.023621
2016-03-13    0.008553
2016-03-14    0.012509
2016-03-15    0.015708
2016-03-16    0.016181
2016-03-17    0.027624
2016-03-18    0.007368
2016-03-19    0.015281
2016-03-20    0.020351
2016-03-21    0.020517
2016-03-22    0.020517
2016-03-23    0.018290
2016-03-24    0.019451
2016-03-25    0.018551
2016-03-26    0.016584
2016-03-27    0.015115
2016-03-28    0.020422
2016-03-29    0.021938
2016-03-30    0.024118
2016-03-31    0.023573
2016-04-01    0.023407
2016-04-02    0.024971
2016-04-03    0.025042
2016-04-04    0.023715
2016-04-05    0.127674
2016-04-06    0.224834
2016-04-07    0.134071
Name: last_seen, dtype: float64

### It seems that opposite to date_crawled 2016-04-06 2016-04-07 had the highest frequencies.

In [67]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).tail(30)

2016-02-25    0.000071
2016-02-14    0.000047
2016-02-18    0.000047
2016-02-24    0.000047
2016-02-12    0.000047
2016-02-20    0.000047
2016-01-27    0.000047
2016-02-02    0.000047
2016-02-26    0.000047
2016-01-10    0.000047
2016-02-05    0.000047
2016-02-11    0.000024
2016-01-07    0.000024
2016-02-07    0.000024
2016-02-22    0.000024
2016-01-13    0.000024
2016-02-01    0.000024
2015-12-05    0.000024
2016-01-14    0.000024
2016-02-17    0.000024
2015-12-30    0.000024
2015-08-10    0.000024
2016-01-16    0.000024
2015-11-10    0.000024
2016-02-16    0.000024
2015-06-11    0.000024
2016-02-08    0.000024
2016-02-09    0.000024
2016-01-03    0.000024
2015-09-09    0.000024
Name: ad_created, dtype: float64

In [68]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).head(30)

2016-04-03    0.039494
2016-03-20    0.038167
2016-03-21    0.037480
2016-04-04    0.036935
2016-03-12    0.036698
2016-04-02    0.035727
2016-03-14    0.035253
2016-03-07    0.035135
2016-03-28    0.034969
2016-04-01    0.034329
2016-03-29    0.033879
2016-03-15    0.033618
2016-03-30    0.033405
2016-03-11    0.033334
2016-03-09    0.033168
2016-03-19    0.033145
2016-03-08    0.033097
2016-03-26    0.032386
2016-03-23    0.032315
2016-03-22    0.032292
2016-03-10    0.032078
2016-03-31    0.031960
2016-03-17    0.031202
2016-03-25    0.031131
2016-03-27    0.030775
2016-03-16    0.029686
2016-03-24    0.029117
2016-03-05    0.022554
2016-03-13    0.017034
2016-03-06    0.015494
Name: ad_created, dtype: float64

### The days are more spread out, that makes sense because the crawler was probably not working all the time but rather specific days. It seems that the beginning and the end of the year is usually a time when people are not very keen to create ads for the cars. People seem more interested in selling starting March, it may be that they have bought a new car for holidays and a few months later realise they don't really want it.

In [69]:
autos['registration_year'].describe()

count    42209.000000
mean      2003.246180
std          6.925527
min       1927.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [70]:
autos['registration_month'].value_counts(normalize=True, dropna=False).sort_index()

1     0.072520
2     0.066858
3     0.113222
4     0.091142
5     0.091118
6     0.096591
7     0.087351
8     0.070743
9     0.076003
10    0.082091
11    0.075932
12    0.076429
Name: registration_month, dtype: float64

### It can be seen here again that most cars are registred in the middle of the year from 3rd month to 7th month

### I will use value_counts to get the most popular brands and then take the top 20 most frequent brands to then add them to a dictionary along with their mean price and another dictionary with their mean mileage

In [78]:
top_brands_mean_price = {}


brands = autos['brand'].value_counts().index.tolist()
for car_brand in brands[:20]:
    select_cars = autos[autos['brand'] == car_brand][['brand','price_dollars']]
    
    mean_price = select_cars.iloc[:,1].mean()
    
    top_brands_mean_price[car_brand] = round(mean_price, 2)
    
    
top_brands_mean_mileage = {}


brands = autos['brand'].value_counts().index.tolist()
for car_brand in brands[:20]:
    select_cars = autos[autos['brand'] == car_brand][['brand','odometer_km']]
    
    mean_mileage = select_cars.iloc[:,1].mean()
    
    top_brands_mean_mileage[car_brand] = round(mean_mileage, 2)

In [79]:
top_brands_mean_price

{'volkswagen': 6039.6,
 'bmw': 8690.2,
 'opel': 3235.09,
 'mercedes_benz': 8911.97,
 'audi': 9812.5,
 'ford': 4391.52,
 'renault': 2670.77,
 'peugeot': 3241.28,
 'fiat': 3031.51,
 'seat': 4609.14,
 'skoda': 6614.66,
 'nissan': 4984.79,
 'mazda': 4424.17,
 'smart': 3700.43,
 'citroen': 3928.46,
 'toyota': 5283.06,
 'hyundai': 5614.23,
 'mini': 10771.73,
 'volvo': 5122.1,
 'sonstige_autos': 27087.65}

In [80]:
top_brands_mean_mileage

{'volkswagen': 128131.97,
 'bmw': 132496.32,
 'opel': 128634.69,
 'mercedes_benz': 130923.35,
 'audi': 128520.45,
 'ford': 123833.68,
 'renault': 127829.9,
 'peugeot': 126120.28,
 'fiat': 115696.68,
 'seat': 121011.6,
 'skoda': 110124.65,
 'nissan': 117096.04,
 'mazda': 122955.27,
 'smart': 98855.52,
 'citroen': 118444.44,
 'toyota': 115432.1,
 'hyundai': 105517.24,
 'mini': 88065.33,
 'volvo': 138632.91,
 'sonstige_autos': 91319.8}

As it can be seen that the most expensive are sonstive_autos (other brands), that might be because of expensive foreign cars that are not as common and fall under sonstige, let's review that 

In [74]:
print(autos[autos['brand'] == 'sonstige_autos'])

              date_crawled                                               name  \
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
140    2016-03-19 19:47:07           Ssangyong_Actyon_SUV_2.0_xdi2wd_55000_km   
152    2016-03-20 12:45:01                                Ssanyong_Rexton_2.7   
175    2016-03-19 15:48:21                                          MG_MGB_GT   
296    2016-03-07 19:43:23                      Dodge_Nitro_4.0_Automatik_R/T   
...                    ...                                                ...   
49391  2016-03-18 00:55:16  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...   
49612  2016-04-02 23:55:49         Opel_Corsa_1.2_Schiebedach_Voll_fahrbereit   
49720  2016-03-31 21:37:36                                          MG_Andere   
49745  2016-03-19 22:50:20                                         ALPINA_B12   
49941  2016-03-15 00:53:25                   Maserati_Ghibli_Diesel_Automatik   

       seller offer_type  p

### As we can see that is correct, we can already see Lamborgnini Gallardo that goes for 109 999 dollars, we can assume now that that is the case

Other than that the most expensive brand is mini at 107711.73 and followed by audi 9812.5, bmw 8690.2, mercedes-benz 8911.97 that would take the most expensive at the top of the list, where as Renault and Opel are least expensive, but under 3300 dollars.

In [94]:
bmp_series = pd.Series(top_brands_mean_price, name = "mean_price")
bmlp_series = pd.Series(top_brands_mean_mileage, name = "mean_mileage")

df = pd.concat([bmp_series, bmlp_series],axis=1)

df

mean_price  mean_mileage
volkswagen         6039.60     128131.97
bmw                8690.20     132496.32
opel               3235.09     128634.69
mercedes_benz      8911.97     130923.35
audi               9812.50     128520.45
ford               4391.52     123833.68
renault            2670.77     127829.90
peugeot            3241.28     126120.28
fiat               3031.51     115696.68
seat               4609.14     121011.60
skoda              6614.66     110124.65
nissan             4984.79     117096.04
mazda              4424.17     122955.27
smart              3700.43      98855.52
citroen            3928.46     118444.44
toyota             5283.06     115432.10
hyundai            5614.23     105517.24
mini              10771.73      88065.33
volvo              5122.10     138632.91
sonstige_autos    27087.65      91319.80